### PR TITLE
Add elliptic curve trait

### PR DESCRIPTION
**Pull request by ajgara**
- Add a generic trait for all the types of curves. (ShortWeierstrass, Edwards, Montgomery, etc...).
- Refactor in PointRepresentation and EllipticCurve representation to make it easier for the users.

### CODE REVIEW

**Review by Oppen — APPROVED** [review]

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs [hunk 1]
```diff
@@ -12,8 +14,28 @@ const BLS12377_MAIN_SUBGROUP_ORDER: U384 =
 /// The description of the curve.
 #[derive(Clone, Debug)]
 pub struct BLS12377Curve;
-impl IsShortWeierstrass for BLS12377Curve {
+
+impl IsEllipticCurve for BLS12377Curve {
     type BaseField = BLS12377PrimeField;
+    type PointRepresentation = ProjectivePoint<Self>;
+
+    fn generator() -> Self::PointRepresentation {
```
Comment (ilitteri) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs:22]: I think that it'd be a good idea for the `generator` function should be `const` in the future. We can file an issue for this.

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs [hunk 2]
```diff
@@ -96,21 +110,21 @@ mod tests {
 
     #[test]
     #[should_panic]
-    fn create_invalid_points_panicks() {
-        ProjectivePoint::<BLS12377Curve>::new([FEE::from(1), FEE::from(1), FEE::from(1)]);
+    fn create_invalid_points_panics() {
```
Comment (ilitteri) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs:113]: We should evaluate if it still makes sense to have this test, given that it'll only panic only in debug idk if it is worth testing.

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs (hunk 1 above)
Comment (ilitteri) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs:22]: Or at least to have a `const` version.

**Review by ilitteri — APPROVED** [review]
LGTM, just left some comments to discuss.

**Review by jrchatruc — APPROVED** [review]

**Comment by schouhy** [comment]
@ilitteri We are working on another PR that includes other refactors as well. We'll add your suggestions in that PR.

**Pull request merged**